File: services/telemetry-processor/app/api/health.py
from fastapi import APIRouter, Response
from sqlalchemy import text
from app.core.database import engine
from datetime import datetime
import logging

logger = logging.getLogger(__name__)
router = APIRouter()


@router.get("/health")
async def health_check():
    """Health check endpoint"""
    try:
        # Check database connection
        async with engine.connect() as conn:
            await conn.execute(text("SELECT 1"))
            
        return {
            "status": "healthy",
            "timestamp": datetime.utcnow().isoformat(),
            "version": "1.0.0",
            "checks": [
                {"name": "database", "status": "up", "message": "Connected"}
            ]
        }
    except Exception as e:
        logger.error(f"Health check failed: {e}")
        return Response(
            status_code=503,
            content={
                "status": "unhealthy",
                "timestamp": datetime.utcnow().isoformat(),
                "error": str(e)
            }
        )


@router.get("/health/ready")
async def readiness_check():
    """Readiness probe for Kubernetes"""
    try:
        async with engine.connect() as conn:
            await conn.execute(text("SELECT 1"))
        return {"ready": True}
    except Exception:
        return Response(status_code=503, content={"ready": False})


@router.get("/health/live")
async def liveness_check():
    """Liveness probe for Kubernetes"""
    return {"alive": True, "timestamp": datetime.utcnow().isoformat()}

File: services/auth-service/src/routes/health.ts
import { Router, Request, Response } from 'express';
import mongoose from 'mongoose';
import { register } from '../utils/metrics';
import { HealthStatus } from '@energix/shared-types';

const router = Router();

// GET /health - Health check endpoint
router.get('/', async (req: Request, res: Response) => {
  const checks: HealthStatus['checks'] = [];

  // Check MongoDB connection
  const dbStatus = mongoose.connection.readyState === 1 ? 'up' : 'down';
  checks.push({
    name: 'mongodb',
    status: dbStatus,
    responseTime: 0,
    message: dbStatus === 'up' ? 'Connected' : 'Disconnected',
  });

  const overallStatus: HealthStatus['status'] = checks.every((c) => c.status === 'up') ? 'healthy' : 'unhealthy';

  const health: HealthStatus = {
    status: overallStatus,
    timestamp: new Date(),
    version: process.env.npm_package_version || '1.0.0',
    checks,
  };

  const statusCode = overallStatus === 'healthy' ? 200 : 503;
  res.status(statusCode).json(health);
});

// GET /health/ready - Readiness probe
router.get('/ready', (req: Request, res: Response) => {
  const isReady = mongoose.connection.readyState === 1;
  res.status(isReady ? 200 : 503).json({
    ready: isReady,
  });
});

// GET /health/live - Liveness probe
router.get('/live', (req: Request, res: Response) => {
  res.json({
    alive: true,
    timestamp: new Date().toISOString(),
  });
});

// GET /metrics - Prometheus metrics
router.get('/metrics', async (req: Request, res: Response) => {
  res.set('Content-Type', register.contentType);
  res.end(await register.metrics());
});

export default router;
